lib/submit-request.js

```diff
@@ -180,6 +183,15 @@ SubmitRequest.prototype.apply = function(callback) {
   this.backend.trigger(this.backend.MIDDLEWARE_ACTIONS.apply, this.agent, this, function(err) {
     if (err) return callback(err);
 
+    /**
+     * Middleware is able to opt in to store old snapshot
+     * which could be useful in cases where we want to compare the old snapshot
+     * (snapshot without the ops applied) and the new snapshot (with the ops applied)
+     */
+    if (request.preservePreapplySnapshot) {
+      request.preapplySnapshot = util.clone(request.snapshot);
```

Comment:
Now that I see this written down, is this any more advantageous than consumers just doing this themselves in middleware directly?

With this approach:

```js
backend.use('apply', (request, next) => {
  request.preservePreapplySnapshot = true;
});
```

Or we can just do this with no changes to ShareDB at all:

```js
backend.use('apply', (request, next) => {
  request.preapplySnapshot = clone(request.snapshot);
});
```

Reply:
Yeah actaully you are right, but there is an argument to be made that it should be part of sharedb itself for simplicity sake, but I am fine to just close this PR and just add some example of this use case in the docs.

Reply:
We decided to just add the documentation about this use case